***STA130 HW01***

Alexander Zubchenko

***1.*** Importing a dataset containing info about Olympic Athletes. - https://chatgpt.com/share/1a1f156c-e488-46de-8918-99794e37b801

In [43]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2021/2021-07-27/olympics.csv"
df = pd.read_csv(url)
df

id                      name sex   age  height  weight  \
0            1                 A Dijiang   M  24.0   180.0    80.0   
1            2                  A Lamusi   M  23.0   170.0    60.0   
2            3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN   
3            4      Edgar Lindenau Aabye   M  34.0     NaN     NaN   
4            5  Christine Jacoba Aaftink   F  21.0   185.0    82.0   
...        ...                       ...  ..   ...     ...     ...   
271111  135569                Andrzej ya   M  29.0   179.0    89.0   
271112  135570                  Piotr ya   M  27.0   176.0    59.0   
271113  135570                  Piotr ya   M  27.0   176.0    59.0   
271114  135571        Tomasz Ireneusz ya   M  30.0   185.0    96.0   
271115  135571        Tomasz Ireneusz ya   M  34.0   185.0    96.0   

                  team  noc        games  year  season            city  \
0                China  CHN  1992 Summer  1992  Summer       Barcelona   
1                China  CHN  2012 Summer  2012  Summer          London   
2              Denmark  DEN  1920 Summer  1920  Summer       Antwerpen   
3       Denmark/Sweden  DEN  1900 Summer  1900  Summer           Paris   
4          Netherlands  NED  1988 Winter  1988  Winter         Calgary   
...                ...  ...          ...   ...     ...             ...   
271111        Poland-1  POL  1976 Winter  1976  Winter       Innsbruck   
271112          Poland  POL  2014 Winter  2014  Winter           Sochi   
271113          Poland  POL  2014 Winter  2014  Winter           Sochi   
271114          Poland  POL  1998 Winter  1998  Winter          Nagano   
271115          Poland  POL  2002 Winter  2002  Winter  Salt Lake City   

                sport                                     event medal  
0          Basketball               Basketball Men's Basketball   NaN  
1                Judo              Judo Men's Extra-Lightweight   NaN  
2            Football                   Football Men's Football   NaN  
3          Tug-Of-War               Tug-Of-War Men's Tug-Of-War  Gold  
4       Speed Skating          Speed Skating Women's 500 metres   NaN  
...               ...                                       ...   ...  
271111           Luge                Luge Mixed (Men)'s Doubles   NaN  
271112    Ski Jumping  Ski Jumping Men's Large Hill, Individual   NaN  
271113    Ski Jumping        Ski Jumping Men's Large Hill, Team   NaN  
271114      Bobsleigh                      Bobsleigh Men's Four   NaN  
271115      Bobsleigh                      Bobsleigh Men's Four   NaN  

[271116 rows x 15 columns]

***1.*** Detecting/Confirming missing values in our dataset

In [44]:
df.isna().sum()

id             0
name           0
sex            0
age         9474
height     60171
weight     62875
team           0
noc            0
games          0
year           0
season         0
city           0
sport          0
event          0
medal     231333
dtype: int64

***2.***
Printing the number of rows & columns in the dataset:
    - https://chatgpt.com/share/44115c01-cd33-4eed-a592-e15a02ad203c

In [45]:
num_rows = df.shape[0] # Making variable for # of rows
num_columns = df.shape[1] # Making a variable for # of columns
print("Rows: " + str(num_rows) + "\nColumns: " + str(num_columns))

Rows: 271116
Columns: 15


In the context of our dataset, the ***observations*** are the rows. They are what we are analyzing. 

The ***variables*** are specific attributes of each observation. In the instance of the dataset above, the Olympics athletes are the observations. Their medals, team, height, weight, etc, are the variables.

***3.***

f.describe() prints out statistics relating to each individual NUMERICAL column. 

In the context of our dataset, we can find the mean, count, std, minimum, etc, relating to variables such as:
    
    - id
    - age
    - height
    - weight
    - participation year


In [46]:
print(df.describe())

                  id            age         height         weight  \
count  271116.000000  261642.000000  210945.000000  208241.000000   
mean    68248.954396      25.556898     175.338970      70.702393   
std     39022.286345       6.393561      10.518462      14.348020   
min         1.000000      10.000000     127.000000      25.000000   
25%     34643.000000      21.000000     168.000000      60.000000   
50%     68205.000000      24.000000     175.000000      70.000000   
75%    102097.250000      28.000000     183.000000      79.000000   
max    135571.000000      97.000000     226.000000     214.000000   

                year  
count  271116.000000  
mean     1978.378480  
std        29.877632  
min      1896.000000  
25%      1960.000000  
50%      1988.000000  
75%      2002.000000  
max      2016.000000  


- For non-numerical elements, we can use the .value_counts() method to find a more detailed summary of a column.


Printing out summary of Athlete Names (Column 1)

In [47]:
print(df['name'].value_counts())

name
Robert Tait McKenzie            58
Heikki Ilmari Savolainen        39
Joseph "Josy" Stoffel           38
Ioannis Theofilakis             36
Takashi Ono                     33
                                ..
Tatyana Vasilyevna Kalmykova     1
Mariya Lvovna Kalmykova          1
Christine Kalmer                 1
Joannis "Jannis" Kalmazidis      1
Valrie Leclerc                   1
Name: count, Length: 134731, dtype: int64



Printing out summary of Athletes' Sex (Column 2)

In [48]:
print(df['sex'].value_counts())

sex
M    196594
F     74522
Name: count, dtype: int64



Printing out summary of Athletes' Team (Column 3)

In [49]:
print(df['team'].value_counts())

team
United States    17847
France           11988
Great Britain    11404
Italy            10260
Germany           9326
                 ...  
Briar                1
Hannover             1
Nan-2                1
Brentina             1
Digby                1
Name: count, Length: 1184, dtype: int64


 
Printing out summary of Athletes' Sport (Column 4)

In [50]:
print(df['sport'].value_counts())

sport
Athletics        38624
Gymnastics       26707
Swimming         23195
Shooting         11448
Cycling          10859
                 ...  
Racquets            12
Jeu De Paume        11
Roque                4
Basque Pelota        2
Aeronautics          1
Name: count, Length: 66, dtype: int64



Printing out summary of Athletes' Medals (Column 5)

In [51]:
print(df['medal'].value_counts())

medal
Gold      13372
Bronze    13295
Silver    13116
Name: count, dtype: int64


***4.***

The dataset being used has: 
- Non numeric variables 
- Missing values in numeric variabes 

df.shape shows the total number of rows and columns -> (271116, 15) meaning 271,116 rows and 15 columns. df.describe() only provides stats for numeric columns. If it shows stats for only 5 columns, it’s because the other 10 columns are non numeric. 

Also, df.describe() counts only non null values in numeric columns. If there are missing values in a column, the count will be lower than the total number of rows. 

This should explain why discrepancies may arise from the information provided by df.shape and df.describe()



***5.***

df.shape is an example of an attribute, it doesn't end with (). Attributes provide info about an object, such as the dimensions of our dataset. df.describe() is a method methods end with (). This shows that it performs an action or function, like summarizing the data in our dataset. In summary, attributes are details about an object, while methods perform actions on it. Attributes are accessed directly, but methods are called with () to execute their function.

***6.***

- https://chatgpt.com/share/c38da79e-159a-43b7-bcd6-4e6f5c0d4ece

Count: The number of non-null values in each column.

Mean: The average value of the data in each column.

Std: The standard deviation, which measures the amount of variation or dispersion of the data.

Min: The smallest value in each column.

25%: The 25th percentile, meaning 25% of the data falls below this value.

50%: The median, or the 50th percentile, where half of the data falls below and half above this value.

75%: The 75th percentile, meaning 75% of the data falls below this value.

Max: The largest value in each column.

df.describe() only works for numeric data because its statistics (mean, std, etc.) need numbers. It ignores missing values, instead of removing them, which is why it still provides results even with missing data. 

Other methods may not handle missing values as easily, so missing data may need to be removed beforehand. This can be done using df.dropna().

***7.***

- https://chatgpt.com/share/8d34895e-ab83-448b-926b-6950c547be1a

***df.dropna()*** can be used to remove rows with missing data. This method is useful when we MUST keep all our columns, but still want to eliminate missing data from rows.

***del df['col']*** can be used to remove columns. This method is especially useful when a column contains too much missing data. So, we can remove it. 

1) We have a dataset where missing values are not concentrated within a specific column. Since all columns are reliable, we can ue df.dropna() to simply remove specific rows with missing values, keeping all the important columns.

2) We have a dataset where ***only*** one column contains many missing values. Since that column is unreliable, we can remove it using del df['col'].

3) If one column exists with many missing values, removing rows first could therefore remove important values from our dataset. Removing columns with del df['col'] cleans up the dataset, making df.dropna() more efficient, only needing to remove the rows with remaining issues.

4) 

In [64]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2021/2021-07-27/olympics.csv"
df = pd.read_csv(url)
print("Before: ")
df

Before: 


id                      name sex   age  height  weight  \
0            1                 A Dijiang   M  24.0   180.0    80.0   
1            2                  A Lamusi   M  23.0   170.0    60.0   
2            3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN   
3            4      Edgar Lindenau Aabye   M  34.0     NaN     NaN   
4            5  Christine Jacoba Aaftink   F  21.0   185.0    82.0   
...        ...                       ...  ..   ...     ...     ...   
271111  135569                Andrzej ya   M  29.0   179.0    89.0   
271112  135570                  Piotr ya   M  27.0   176.0    59.0   
271113  135570                  Piotr ya   M  27.0   176.0    59.0   
271114  135571        Tomasz Ireneusz ya   M  30.0   185.0    96.0   
271115  135571        Tomasz Ireneusz ya   M  34.0   185.0    96.0   

                  team  noc        games  year  season            city  \
0                China  CHN  1992 Summer  1992  Summer       Barcelona   
1                China  CHN  2012 Summer  2012  Summer          London   
2              Denmark  DEN  1920 Summer  1920  Summer       Antwerpen   
3       Denmark/Sweden  DEN  1900 Summer  1900  Summer           Paris   
4          Netherlands  NED  1988 Winter  1988  Winter         Calgary   
...                ...  ...          ...   ...     ...             ...   
271111        Poland-1  POL  1976 Winter  1976  Winter       Innsbruck   
271112          Poland  POL  2014 Winter  2014  Winter           Sochi   
271113          Poland  POL  2014 Winter  2014  Winter           Sochi   
271114          Poland  POL  1998 Winter  1998  Winter          Nagano   
271115          Poland  POL  2002 Winter  2002  Winter  Salt Lake City   

                sport                                     event medal  
0          Basketball               Basketball Men's Basketball   NaN  
1                Judo              Judo Men's Extra-Lightweight   NaN  
2            Football                   Football Men's Football   NaN  
3          Tug-Of-War               Tug-Of-War Men's Tug-Of-War  Gold  
4       Speed Skating          Speed Skating Women's 500 metres   NaN  
...               ...                                       ...   ...  
271111           Luge                Luge Mixed (Men)'s Doubles   NaN  
271112    Ski Jumping  Ski Jumping Men's Large Hill, Individual   NaN  
271113    Ski Jumping        Ski Jumping Men's Large Hill, Team   NaN  
271114      Bobsleigh                      Bobsleigh Men's Four   NaN  
271115      Bobsleigh                      Bobsleigh Men's Four   NaN  

[271116 rows x 15 columns]

In [66]:
del df['medal']
del df['weight']
del df['age']
print("After: ")
df.dropna()

After: 


id                      name sex  height         team  noc  \
0            1                 A Dijiang   M   180.0        China  CHN   
1            2                  A Lamusi   M   170.0        China  CHN   
4            5  Christine Jacoba Aaftink   F   185.0  Netherlands  NED   
5            5  Christine Jacoba Aaftink   F   185.0  Netherlands  NED   
6            5  Christine Jacoba Aaftink   F   185.0  Netherlands  NED   
...        ...                       ...  ..     ...          ...  ...   
271111  135569                Andrzej ya   M   179.0     Poland-1  POL   
271112  135570                  Piotr ya   M   176.0       Poland  POL   
271113  135570                  Piotr ya   M   176.0       Poland  POL   
271114  135571        Tomasz Ireneusz ya   M   185.0       Poland  POL   
271115  135571        Tomasz Ireneusz ya   M   185.0       Poland  POL   

              games  year  season            city          sport  \
0       1992 Summer  1992  Summer       Barcelona     Basketball   
1       2012 Summer  2012  Summer          London           Judo   
4       1988 Winter  1988  Winter         Calgary  Speed Skating   
5       1988 Winter  1988  Winter         Calgary  Speed Skating   
6       1992 Winter  1992  Winter     Albertville  Speed Skating   
...             ...   ...     ...             ...            ...   
271111  1976 Winter  1976  Winter       Innsbruck           Luge   
271112  2014 Winter  2014  Winter           Sochi    Ski Jumping   
271113  2014 Winter  2014  Winter           Sochi    Ski Jumping   
271114  1998 Winter  1998  Winter          Nagano      Bobsleigh   
271115  2002 Winter  2002  Winter  Salt Lake City      Bobsleigh   

                                           event  
0                    Basketball Men's Basketball  
1                   Judo Men's Extra-Lightweight  
4               Speed Skating Women's 500 metres  
5             Speed Skating Women's 1,000 metres  
6               Speed Skating Women's 500 metres  
...                                          ...  
271111                Luge Mixed (Men)'s Doubles  
271112  Ski Jumping Men's Large Hill, Individual  
271113        Ski Jumping Men's Large Hill, Team  
271114                      Bobsleigh Men's Four  
271115                      Bobsleigh Men's Four  

[210945 rows x 12 columns]

In [65]:
df.isna().sum()

id             0
name           0
sex            0
age         9474
height     60171
weight     62875
team           0
noc            0
games          0
year           0
season         0
city           0
sport          0
event          0
medal     231333
dtype: int64

After running df.isna().sum(), I concluded that the columns: Height, Age, Weight, and Medal contained a high number of missing values. To remove the unreliable columns, I ran del df['col']. Next, I used df.dropna() to remove all rows with missing values.

Through doing this, I could remove unreliable columns from my dataset, while also keeping as much reliable data as possible.

***8.***

1) - https://chatgpt.com/share/eea3fd72-3b9e-4dd5-9cf8-f6ffcfb8feeb
    
    The method df.groupby("col1")["col2"].describe() creates a subset of the dataset for each unique value in col1.
    Next, for each group, statistics will be provided for col2: count, mean, standard deviation, minimum, 25th percentile, median (50th percentile), 75th percentile, and maximum.
        
        Ex.

In [73]:
#Example

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.groupby("class")["fare"].describe())

        count       mean        std  min       25%      50%   75%       max
class                                                                      
First   216.0  84.154687  78.380373  0.0  30.92395  60.2875  93.5  512.3292
Second  184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
Third   491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500


2. - https://chatgpt.com/share/eea3fd72-3b9e-4dd5-9cf8-f6ffcfb8feeb

df.describe() will count the total number of non-missing values in an ***individual column***. 

However, df.groupby("col1")["col2"].describe() will count how many non-missing values of col2 exist within each unique value of ***col1***. 

In summary, the df.describe() gives a count per column, while df.groupby("col1")["col2"] gives a count of how data in col2 is distributed across categories or groups in col1.

3. 

In [1]:
# Error A: Not importing Pandas

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

NameError: name 'pd' is not defined

ChatGPT solution - https://chatgpt.com/share/e7b32a03-a376-433d-b92f-2325df9f13ba

When consulting ChatGPT with my error, I recieved an immediate working solution, informing me that I forgot to import the pandas library. When consulting google with my error, I found several stackoverflow posts regarding similar pandas related issues, but nothing that provided an immediate fix to my problem. In summary, I noticed ChatGPT had a more streamlined fix and helped me resolve my problem quicker.

In [3]:
# Error B: Mistype "titanic.csv" as "titanics.csv"

import pandas as pd
url = "titanics.csv"
df = pd.read_csv(url)

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

ChatGPT solution - https://chatgpt.com/share/a2c785f9-a4c5-4c9e-a978-ca3651c21e0e

When consulting ChatGPT, I recieved a concise fix to my issue. First, I was instructed to double check the path to my file location (if I was loading it locally). Next, I was suggested to check the spelling of my file name, and if using an online source, use an online dataset url. Because the true issue was from the spelling of my filename & the improper url format, I think ChatGPT did a great job of solving my problem.

Attempting to find a fix with a Google search didn't work out so well. Rather than finding an immediate fix, I was suggested unrelated content from various coding forums which didn't properly address my issue. 

In [5]:
# Error C: Try to use a dataframe before it's been assigned into the variable

import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

DF.describe()

NameError: name 'DF' is not defined

ChatGPT solution - https://chatgpt.com/share/d99ce0b1-3501-4953-84d1-07800f5c0ee3
    
The ChatGPT solution correctly pointed out that Python is case sensitive and that I incorrectly called "DF" instead of "df". To fix my error, it suggested to change DF.describe() to df.describe().

When consulting Google, I recieved a similar solution, suggesting that case sensitivity is important in Python, and that I calling the variable "df" is not the same as "DF". Overall, I found both solutions quick and informative.

In [6]:
# Error D: Forget a parantheses

import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url

SyntaxError: incomplete input (2020537918.py, line 5)

ChatGPT solution - https://chatgpt.com/share/da505af9-e80a-4b79-a2b8-d60ccf03ce45

ChatGPT correctly points out that I had missed a closing parantheses. The fix is quick and easy.

Although a Google search correctly points out what a SyntaxError is, it cannot pinpoint where in my code the SyntaxError really is. Therefore, it would be hard for someone to understand where to look to solve their issue. ChatGPT can locate the exact position of the syntax error, and explain how to fix it.

In [12]:
# Error E: Mistype one of the names of the chained functions with the code

import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.group_by("class")["fare"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

ChatGPT solution - https://chatgpt.com/share/f5465673-de4c-4691-9a6e-c7bf2121de5e
    
ChatGPT explains that the method "group_by" doesn't exist in pandas. It explains how the correct method is "groupby", and provides a working fix code snippet.

Surprisingly, Google searching this error provided a link with an identical fix to ChatGPT (Reddit thread explaining the incorrect spelling of "group_by"). In this case, both searching through Google and using a Chatbot provided equally effective solutions.

In [14]:
# Error F: Mistype one of the names of the chained functions with the code

import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

In [17]:
df.groupby("sex")["Age"].describe()

KeyError: 'Column not found: Age'

ChatGPT solution - https://chatgpt.com/share/90d721d1-0ac8-4f1e-8dfe-2b8737290ba7

Unlike earlier examples, ChatGPT encountered some problems finding a solution. When prompted to find a fix, ChatGPT first incorrectly assumed to change "age" to "Age" to match the capital in "Sex". However, after providing ChatGPT with more information from the error response, it correctly identified the solution to revert both to lowercase.

After solving the first instance with "Sex", it seems ChatGPT knew that the columns should be in lowercase, and correctly found the same fix for the reverse version of "Age".

A Google search of my error resulted in no relevant solutions, as this issue was unfortunately far too specific to yield any direct solutions.

In [18]:
df.groupby(sex)["age"].describe()

NameError: name 'sex' is not defined

In [20]:
df.groupby("sex")[age].describe()

NameError: name 'age' is not defined

ChatGPT solution - https://chatgpt.com/share/31551cac-daf2-41a4-855d-becfe6ed890f
    
ChatGPT correctly identified the problem of the column names "sex" and "age" being undefined. The fix suggested was to enclose them in quotations so that they can be read as Strings. 

When searching the error on Google, I found several fixes for related issues with variables being undefined. With enough time, I could definitely trace such findings to a fix. However, ChatGPT is much faster and reliable at pointing out the problem.

***9.***

Yes, I have reviewed the Course WIKI